### PR TITLE
[Form] deprecate submitting non HTML5 formatted dates

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/pull/28712#pullrequestreview-161542293
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by xabbuh** [comment]
needs to be rebased once #28712 is merged up to master

**Comment by nicolas-grekas** [comment]
I'm really unsure about this. Having an option means asking devs to make a choice, and making a choice is hard. Here, enabling strict mode (Yay, strict everything!) means lowering the compatibility with the ecosystem. That's contra the robustness principle.
We have a native date parser that is excellent at parsing real-world date formats, there is no better option for me. We should leverage it.

**Comment by xabbuh** [comment]
To me, this goes along with #28753. If one wants to use HTML 5 features, they should follow the standards IMO. Custom formats are okay, but then you need to be explicit about this (i.e. set `html5` to `false` and configure the `format` option explicitly). Without being strict, there is too much room for guessing from my point of view which unnecessarily adds uncertainty that makes it hard to fix any date (and time) related bugs.

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeToHtml5LocalDateTimeTransformer.php [hunk 1]
```diff
@@ -22,6 +22,23 @@ class DateTimeToHtml5LocalDateTimeTransformer extends BaseDateTimeTransformer
...
     const HTML5_FORMAT = 'Y-m-d\\TH:i:s';
 
+    private $strict;
+
+    public function __construct(string $inputTimezone = null, string $outputTimezone = null)
+    {
+        parent::__construct($inputTimezone, $outputTimezone);
+
+        if (2 < \func_num_args()) {
+            $this->strict = \func_get_arg(2);
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeToHtml5LocalDateTimeTransformer.php]: no need to to use `func_get_args`. Adding an optional argument in a constructor is not a BC break (a child constructor does not need to have a compatible signature)

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeToHtml5LocalDateTimeTransformer.php [hunk 2]
```diff
@@ -81,7 +98,15 @@ public function reverseTransform($dateTimeLocal)
             return;
         }
 
-        if (!preg_match('/^(\d{4})-(\d{2})-(\d{2})[T ]\d{2}:\d{2}(?::\d{2})?$/', $dateTimeLocal, $matches)) {
+        // to maintain backwards compatibility we do not strictly validate the submitted date
+        // see https://github.com/symfony/symfony/issues/28699
+        if (!$this->strict) {
+            $regex = '/^(\d{4})-(\d{2})-(\d{2})[T ]\d{2}:\d{2}(?::\d{2})?/';
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeToHtml5LocalDateTimeTransformer.php]: to maintain BC, shouldn't it use the old transformer instead (which did not have that regex either, so you are only partially fixing the BC break)

**Comment by stof** [comment]
@nicolas-grekas if you have a form type configured as HTML5_FORMAT, supporting other formats looks weird.

The option is here to opt-in for the new behavior at the time you want (needed in case your JS currently generates a different format for instance)

**Comment by stof** [comment]
We may also want to harden DateTimeToRfc3339Transformer to actually support only RFC 3339 (using `DateTime::createFromFormat`)

**Comment by nicolas-grekas** [comment]
I'm :-1: here: data transformers are not where strictness should apply. Strictness is for validators.
My reasoning is "why?" should we make this change? What would it improve? On the maintainer side, nothing. On the user side, it would just make using the HTTP endpoint less compatible with UI widgets into the wild, thus making things harder to use actually.
Same as why XML didn't win the race and we're using loose HTML5 parser for the benefit of the robustness of the Web.

**Comment by nicolas-grekas** [comment]
@xabbuh ok with my main objection: not worth adding a new option as DX may suffer?

**Comment by xabbuh** [comment]
I am closing here for now. We can consider to reopen when we notice that a less strict date format handling leads to avoidable bugs.

**Pull request closed without merging**